File: 538.Convert-BST-to-Greater-Tree.java
// https://leetcode.com/problems/convert-bst-to-greater-tree/
//
// algorithms
// Easy (50.39%)
// Total Accepted:    73,388
// Total Submissions: 145,653
// beats 100.0% of java submissions


/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    
    static int curSum = 0;
    
    public TreeNode convertBST(TreeNode root) {
        if (root == null) {
            return root;
        }
        
        Solution.curSum = 0;
        recursive(root);
        
        return root;
    }
    
    public void recursive(TreeNode root) {
        if (root == null) {
            return;
        }
        
        this.recursive(root.right);
        root.val = root.val + Solution.curSum;
        Solution.curSum = root.val;
        this.recursive(root.left);
    }
}

File: 538.Convert-BST-to-Greater-Tree.py
# https://leetcode.com/problems/convert-bst-to-greater-tree/
#
# algorithms
# Easy (51.97%)
# Total Accepted:    88,918
# Total Submissions: 171,108
# beats 100.0% of python submissions


# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def convertBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        acc = [0]

        def recursive(node):
            if not node:
                return

            recursive(node.right)

            node.val += acc[0]
            acc[0] = node.val

            recursive(node.left)

        recursive(root)

        return root
